### PR TITLE
DAOS-15411 test: default test-tag based on files modified

### PR DESCRIPTION
**Pull request by daltonbohning**
Skip-build: true
Skip-test: true
Skip-tests: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/tags.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  src/object/: object
+  src/pool/: pool
+  src/rebuild/: rebuild
+  src/security/: security
+  src/tests/suite/daos.*: daos_test
+  src/tests/suite/dfs.*: dfs_test
+  src/tests/suite/dfuse.*: dfuse_test
+  src/tests/ftest*:
+    handler: FtestTagMap
+    exclusive: True
```
Comment (daltonbohning) [inline src/tests/ftest/tags.yaml:31]: stop_on_match?

File: src/tests/ftest/tags.py [hunk 2]
```diff
@@ -87,43 +89,43 @@ def unique_tags(self, exclude=None):
...
-                    if globally_unique_tags and globally_unique_tags.issubset(tags):
-                        file_recommended.append(globally_unique_tags)
-                        continue
-                    # Fallback to just using all of this test's tags
-                    file_recommended.append(tags)
+        classes = self.__mapping[file_path]
+        # Keep track of recommended tags for each method
+        file_recommended = []
+        for class_name, functions in classes.items():
+            for function_name, tags in functions.items():
```
Comment (daltonbohning) [inline src/tests/ftest/tags.py:114]: Use temp list of tags and only add if found. I.e. to support multiple exclusive

File: Jenkinsfile [hunk 3]
```diff
@@ -135,6 +135,44 @@ void rpm_test_post(String stage_name, String node) {
     job_status_update()
 }
 
+/**
+ * Update env.pragmas with default commit pragmas if not set.
+ */
+Map update_default_commit_pragmas() {
```
Comment (daltonbohning) [inline Jenkinsfile:141]: print message to say whether using user tags or auto

File: Jenkinsfile [hunk 4]
```diff
@@ -135,6 +135,44 @@ void rpm_test_post(String stage_name, String node) {
...
+    print("env.pragmas before modification:")
+    print(env.pragmas)
+    Map env_pragmas = [:]
+    if (env.pragmas) {
+        env_pragmas = "${env.pragmas}"[1..-2].split(', ').collectEntries { entry ->
+            String[] pair = entry.split('= ')
+            [(pair.first()): pair.last()]
+        }
+    }
```
Comment (daltonbohning) [inline Jenkinsfile:150]: librarify to pipelinelib

**Comment by daltonbohning** [comment]
Run showing auto-generated `Test-tag: DfuseMUPerms DmvrDstCreate pr`:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13964/53/pipeline

**Comment by daltonbohning** [comment]
This is ready for final-ish review now. From a high level, the flow is:
- Jenkins `Get Commit Message` stage calls `gen_commit_pragmas.py`
  - `gen_commit_pragmas.py` passes a list of files modified to `ftest/tags.py:files_to_tags`
    - `files_to_tags` reads `ftest/tags.yaml` for the mapping and uses special handling for ftest files
    - `files_to_tags` returns a set of avocado tags corresponding to the files
  - `gen_commit_pragmas.py` uses the tags to generate `Test-tag: <tags>`
  - `gen_commit_pragmas.py` prints the pragma(s)
- Jenkins passes the pragmas output to `updatePragmas` and lets pipeline-lib handle them as it does other pragmas

A user can use `Test-tag` in their commit to override this auto-generated list, or they can use `Features` to add to this auto-generated list.

The reason for hijacking commit pragmas like this is because it's the simplest control we have for configuring pipeline behavior. In  the future, it would be useful to enable/disable other stages in `gen_commit_pragmas.py`.

**Comment by daltonbohning** [comment]
@brianjmurrell I imagine we will want similar behavior for the GHA runs. Since the mapping only works for ftest files currently, I'd like to let this mature some before adding support in the GHA

**Review by phender — APPROVED** [review]

**Review by shimizukko — APPROVED** [review]

**Comment by daltonbohning** [comment]
Run showing auto-generated Test-tag: DfuseMUPerms DmvrDstCreate pr:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13964/53/pipeline

Latest run is a clean pr-only

**Comment by daltonbohning** [comment]
Please use this description for merge
```
- Automatically include test tags for specific functional tests modified.
- Add gen_commit_pragmas.py which can be expanded in the future to
  automatically set commit pragmas based on files modified.
- Add tags.yaml which can be expanded in the future to include additional
  files -> feature tag mappings.
```

**Pull request merged**